feat: Make the communicate_with_quorum grace period configurable

huitseeker opened the pull request:
- Renamed `GRACE_PERIOD` to `DEFAULT_GRACE_PERIOD` ~~and altered its value~~,
- added grace period argument in communicate with quorum,
- added that value to `ChainClientOptions` and threaded through the `Client` builder,
- Refactored the `end_time` setting logic in the `communicate_with_quorum` function to improve the handling of unreachable quorum on any key,
- ~~also corrects a handful of typos (in comments and error messages) as I was reading in the area~~

Fixes #2836. /cc @afck for review @Twey for context

ma2bd commented on linera-core/src/updater.rs:
```diff
@@ -31,9 +31,9 @@ use crate::{
-const GRACE_PERIOD: f64 = 0.2;
+/// The default amount of time we wait for additional validators to contribute 
+/// to the result, as a fraction of how long it took to reach a quorum.
+const DEFAULT_GRACE_PERIOD: f64 = 0.1;
```
Please keep the current value

huitseeker replied:
The value was changed following the advice in #2836.

afck replied:
I don't have a strong opinion; sorry for the confusion!
(The idea was that since `0.2` means we make the _happy path_ 20% slower it improves user experience to reduce this. And we will need additional tools to help slow validators catch up anyway. But we can always change the default later.)

ma2bd commented on linera-core/src/updater.rs:
```diff
@@ -97,14 +97,16 @@ pub enum CommunicationError<E: fmt::Debug> {
     group_by: G,
     execute: F,
+    // Grace period as a fraction of time taken to reach quorum
+    grace_period: impl Into<Option<f64>>,
```
We generally avoid `impl Into` (to keep type-checking strong and the code normalized).

huitseeker replied:
reverted to `Option<f64>`

ma2bd commented on linera-core/src/updater.rs:
```diff
@@ -167,13 +170,15 @@ where
-            end_time = Some(Instant::now() + start_time.elapsed().mul_f64(GRACE_PERIOD));
+        // If it becomes clear that no key can reach a quorum, break early.
+        if highest_key_score + remaining_votes < committee.quorum_threshold() {
+            break 'vote_wait;
```
cc @afck for reviewing this part

afck replied:
Makes sense to me. :+1: 
(An alternative would be to set `end_time` to `now()`, but maybe the `break` is cleaner.)

ma2bd commented:
Thanks @huitseeker for the PR!
* One cannot change the protocol behavior (here default value) in a PR called "refactor"
* Consider using several distinct commits in the PR next time :) to separate the easy stuff from the more serious refactoring
* Seems like you need to call `cargo +nightly fmt`

huitseeker commented:
- removed typo fixes,
- split change of DEFAULT_GRACE_PERIOD to another commit.

huitseeker commented:
> One cannot change the protocol behavior (here default value) in a PR called "refactor" (also I think 0.2 is fine for now)

Ah, yeah, i had not noticed this was public. I would actually advise switching this low level function to `pub(crate)`

afck reviewed (approved):
Looks good to me, but it doesn't actually make it configurable yet, i.e. add a config option.

huitseeker commented:
Thanks for the feedback @ma2bd and @afck. I have:
- removed the commit that changed the default value of the grace period (now at the original value of 0.2)
- added a commit that includes the `grace_period` parameter as part of the `ChainClientOptions` and threads that value through the `Client` builder.

afck commented on linera-core/src/client/mod.rs:
```diff
@@ -157,6 +157,9 @@ where
     message_policy: MessagePolicy,
     /// Whether to block on cross-chain message delivery.
     cross_chain_message_delivery: CrossChainMessageDelivery,
+    /// Grace period for communicate_with_quorum as a fraction of time taken to reach quorum.
+    /// If None, uses the default value.
```
```suggestion
    /// Grace period for `communicate_with_quorum` as a fraction of time taken to reach quorum.
    /// If `None`, uses the default value.
```

Or maybe we should not mention the name of the method. "An additional delay, after reaching a quorum, to wait for additional validator signatures."?

afck reviewed (commented):
Thanks; but that still doesn't make it user-configurable: It doesn't add a CLI option yet. (I think that needs to be added to `client_options.rs`.)

afck commented on linera-core/src/client/mod.rs:
```diff
@@ -1516,6 +1527,7 @@ where
                         .await
                 })
             },
+            None,
```
Did you mean to use `self.options.grace_period`, or is this deliberately `None`?

afck commented on linera-client/src/client_options.rs:
```diff
@@ -156,6 +156,11 @@ pub struct ClientOptions {
+    /// An additional delay, after reaching a quorum, to wait for additional validator signatures,
+    /// as a fraction of time taken to reach quorum. If `None`, uses the default value.
+    #[arg(long)]
+    pub grace_period: Option<f64>,
```
(We could also make this a percentage. Not sure if users will find it more intuitive to specify `--grace-period 20` or `--grace-period 0.2`… or `--grace-period 20%`? Anyway, we can always change this later.)

ma2bd replied:
please no percentage. We don't need to create footguns

afck commented on linera-client/src/client_options.rs:
```diff
@@ -156,6 +156,11 @@ pub struct ClientOptions {
     pub restrict_chain_ids_to: Option<HashSet<ChainId>>,
+
+    /// An additional delay, after reaching a quorum, to wait for additional validator signatures,
+    /// as a fraction of time taken to reach quorum. If `None`, uses the default value.
```
Since this is user-visible maybe we shouldn't mention Rust values like `None`. ("If omitted"?)

afck reviewed (approved):
No blockers from my side!

ma2bd commented on linera-core/src/client/mod.rs:
```diff
@@ -157,6 +157,9 @@ where
     cross_chain_message_delivery: CrossChainMessageDelivery,
+    /// An additional delay, after reaching a quorum, to wait for additional validator signatures,
+    /// as a fraction of time taken to reach quorum. If `None`, uses the default value.
+    grace_period: Option<f64>,
```
Note that we could remove the `Option` entirely by injecting the default value during CLI parsing (with or without moving the constant).

huitseeker replied:
That would work for the CLI.  Clients created in `linera-service` would also need to inject the default.
 
I've added a commit that shows what that looks like. This `linera-service` concern also shows why it may make sense to keep `DEFAULT_GRACE_PERIOD` in the core crate.

ma2bd reviewed (approved):
Thanks!
